src/lib/security/selinux_config.rb

```diff
@@ -25,92 +25,156 @@ module Security
+    # security module. Instead, it just fallback to the kernel configuration at the compile time,
+    # which in SUSE is to use AppArmor according to the CONFIG_LSM variable. So, it could be said
+    # that dropping all policy param to disabling SELinux is safe enough.
+    #
+    # To know more, please visit the LSM Usage documentation at
+    # https://www.kernel.org/doc/html/latest/admin-guide/LSM/index.html
+    # and/or grep for CONFIG_LSM in /boot/config-*
     POLICIES = {
+      # NOTE: setting the "security" module to :missing, which means to remove the options from the
+      #       kernel
       disabled:   { "security" => :missing, "selinux" => :missing, "enforcing" => :missing },
       permissive: { "security" => "selinux", "selinux" => "1", "enforcing" => :missing },
       enforcing:  { "security" => "selinux", "selinux" => "1", "enforcing" => "1" },
     }.freeze
     private_constant :POLICIES
 
-    def configured_policy
-      match_policy(policy_options_from_kernel) || :disabled
+    # Returns the requested policy, if exists
+    #
+    # @param key [String, Symbol] the policy identifier
+    # @return [Hash, nil] options for matched policy or nil if none
+    def find_policy(key)
+      POLICIES[key.to_sym]
     end
 
-    # Propose a policy according to the value set in the control-file
+    # Proposes a policy based on `selinux_policy` value set in the control file
     def propose_policy
-      key = :enforcing # read it from control-file
+      policy_key = Yast::ProductFeatures.GetFeature("globals", "selinux_policy").to_sym
 
-      log.info "Proposing the `#{key}` SELinux policy"
-      update_policy(policy_key)
+      log.info "Proposing the `#{policy_key}` SELinux policy"
+      self.policy = policy_key
       save
     end
 
-    def find_policy(key)
-      POLICIES[key]
-    end
+    # Returns the configured SELinux policy according to params in kernel command line
+    #
+    # @see #policy_from_kernel_params
+    #
+    # @return [Symbol] the policy identifier
+    def configured_policy
+      options = policy_from_kernel_params
+
+      return :disabled if options.empty?
+
+      security_module = options["security"]
+      module_disabled = options["selinux"].to_i <= 0
+      enforcing_mode  = options["enforcing"].to_i > 0
+
+      return :disabled if security_module != "selinux" || module_disabled
+      return :permissive unless enforcing_mode
 
-    def match_policy(policy_options)
-      POLICIES.key(policy_options)
+      :enforcing
     end
 
-    def policy_options_from_kernel
-      Hash[*POLICY_KEYS.flat_map { |key| [key, read_param(key)] }]
+    # Returns the SELinux configuration based on options set in the kernel command line
+    #
+    # @return [Symbol] the policy identifier
+    def policy_from_kernel_params
+      options = Hash[*POLICY_KEYS.flat_map { |key| [key, read_param(key)] }]
+      options.filter { |_, value| value != :missing }
```

Comment:
still this code does not address if there are multiple entries on cmdline. E.g. consider `security=selinux selinux=0 enforcing=0 selinux=1 enforcing=1`

Reply:
Right.

But It's a little bit crazy to know how to proceed in such cases. See my tests below

#### `security=selinux selinux=0 enforcing=0 selinux=1 enforcing=1`

SELinux Permissive wins

```
localhost:~ # cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.10.1-1-default root=UUID=1d930088-0e08-4013-b2fe-5e639eddf91a splash=silent quiet security=selinux selinux=0 enforcing=0 selinux=1 enforcing=1
localhost:~ # getenforce
Permissive
```

#### `security=selinux selinux=0 enforcing=1 selinux=1 enforcing=0`

SELinux Enforcing wins

```
localhost:~ # cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.10.1-1-default root=UUID=1d930088-0e08-4013-b2fe-5e639eddf91a splash=silent quiet security=selinux selinux=0 enforcing=1 selinux=1 enforcing=0
localhost:~ # getenforce
Enforcing
```


#### `security=selinux selinux=1 enforcing=0 selinux=0 enforcing=1`

SELinux Disabled wins

```
localhost:~ # cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.10.1-1-default root=UUID=1d930088-0e08-4013-b2fe-5e639eddf91a splash=silent quiet security=selinux selinux=1 enforcing=0 selinux=0 enforcing=1
localhost:~ # getenforce
Disabled
```

Reply:
yeah, that looks really crazy..hopefully bootloader API returns array in order, so it is solvable to implement such logic, I am just not sure how much backward compatible it is Probably the most intersting question for runtime module.

Reply:
> hopefully bootloader API returns array in order

That's what I thought and what I was testing, but seems that it returns the array in the same order,  not in the _right_ one. See attached screenshots

![Screenshot_opensusetumbleweed_2021-02-02_12:15:37](https://user-images.githubusercontent.com/1691872/106602451-ec3cd500-6554-11eb-8aba-b64c42b9b619.png)

![Screenshot_opensusetumbleweed_2021-02-02_12:19:12](https://user-images.githubusercontent.com/1691872/106602457-f068f280-6554-11eb-8332-10a8e111df84.png)

![Screenshot_opensusetumbleweed_2021-02-02_12:24:31](https://user-images.githubusercontent.com/1691872/106602472-f6f76a00-6554-11eb-8331-d71e46345611.png)

Reply:
yes, it does not have such knowledge, so it returns order as it is on cmdline. It is up to you to decide which value wins or if you need to merge it ( as e.g. console means that each value is another device to send output ).

Reply:
My main objection here is that you use `to_i` on result of that get_param which crash if there are more keys

Reply:
Ok, as we agreed in a video call, lets go for the latest value for now.

Thanks Josef.